### PR TITLE
tidb_query: Migrate `json_object` and `json_merge` from builtin_json.rs to tidb_query

### PR DESCRIPTION
**Pull request by mapleFU**
## What have you changed? (mandatory)

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:
* Migrate implementation for `json_array`, `json_object`, `json_merge` from builtin_json.rs to the new impl_json.rs. I migrate 3 functions in one pr because they are similar.
* Add support for `JsonObjectSig` `JsonMergeSig` in rpn_expr/mod.rs for `components/tidb_query`.
* This pr's `JsonObjectSig` is based on #5282 ,please merge #5282 before merge this pr.

This pr require another pr to be merged first.

Please **NOTE** that:
- Do not assume reviewers understand the original issue

## What are the type of changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- New feature (Support new RPN Function).

## How has this PR been tested? (mandatory)

New unit tests.

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect `tidb-ansible` update? (mandatory)

No.

### CODE REVIEW

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 1]
```diff
@@ -29,20 +29,27 @@ fn json_array(args: &[&Option<Json>]) -> Result<Option<Json>> {
...
         let key = match key {
             // json_object should raise an error if key is None(NULL)
-            None => unimplemented!(),
-            Some(v) => String::from_utf8(v.to_owned()).map_err(|e| crate::codec::Error::from(e))?,
-        };
+            None => Err(other_err!(
+                "Data truncation: JSON documents may not contain NULL member names."
+            )),
+            Some(v) => {
+                String::from_utf8(v.to_owned()).map_err(|e| crate::codec::Error::from(e).into())
```
Comment (iosmanthus) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: Could we omit `.map_err(|e| crate::codec::Error::from(e).into())`?

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 2]
```diff
@@ -14,6 +16,83 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
...
+        };
+        pairs.insert(key, value);
+    }
+    Ok(Some(Json::Object(pairs)))
+}
+
+macro_rules! parse_opt {
+    ($expr:expr) => {{
+        match $expr {
+            Some(ref v) => v.to_owned().clone(),
```
Comment (iosmanthus) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: Why we need `.clone()`?

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 3]
```diff
@@ -14,6 +16,83 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
...
+
+/// Required args like `&[(&Option<Byte>, &Option<Json>)]`.
+#[rpn_fn(raw_varg, extra_validator = json_validator)]
+#[inline]
+fn json_object(raw_args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    let mut pairs = BTreeMap::new();
+    for chunk in raw_args.chunks(2) {
+        // chunk.len() must be 1 or 2 here.
+        let key: &Option<Bytes> = Evaluable::borrow_scalar_value_ref(&chunk[0]);
+        let key = match key {
```
Comment (iosmanthus) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: Maybe we could use `ok_or_else` and some combinators here.

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 4]
```diff
@@ -57,6 +136,127 @@ mod tests {
...
+
+        for (vargs, expected) in cases {
+            let vargs = vargs
+                .into_iter()
+                .map(|(key, value)| (Bytes::from(key), value.map(|s| Json::from_str(s).unwrap())))
+                .collect::<Vec<_>>();
+
+            let mut new_vargs: Vec<ScalarValue> = vec![];
+            for (key, value) in vargs.into_iter() {
+                new_vargs.push(ScalarValue::from(key.clone()));
```
Comment (iosmanthus) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: `(key, value)` seems just being used in the loop just one time. Could we remove `.clone()`?

File: components/tidb_query/src/rpn_expr/types/test_util.rs [hunk 5]
```diff
@@ -98,7 +99,19 @@ impl RpnFnScalarEvaluator {
                 ed
             })
             .collect();
-        let func = super::super::map_pb_sig_to_rpn_func(sig, &children_ed).unwrap();
+
+        // use validator_ptr to testing the test arguments.
+        let func: RpnFnMeta = super::super::map_pb_sig_to_rpn_func(sig, &children_ed).unwrap();
```
Comment (iosmanthus) [inline components/tidb_query/src/rpn_expr/types/test_util.rs]: Please merge the https://github.com/tikv/tikv/pull/5282 while it's merged into master.

File: components/tidb_query/src/rpn_expr/impl_json.rs (hunk 4 above)
Comment (mapleFU) [reply components/tidb_query/src/rpn_expr/impl_json.rs]: > `(key, value)` seems just being used in the loop just one time. Could we remove `.clone()`?

Yes, I'll change it.

File: components/tidb_query/src/rpn_expr/impl_json.rs (hunk 1 above)
Comment (mapleFU) [reply components/tidb_query/src/rpn_expr/impl_json.rs]: > Could we omit `.map_err(|e| crate::codec::Error::from(e).into())`?

This expression returns `tidb_query::Error`, do you have any idea to omit it?

File: components/tidb_query/src/rpn_expr/impl_json.rs (hunk 2 above)
Comment (mapleFU) [reply components/tidb_query/src/rpn_expr/impl_json.rs]: > Why we need `.clone()`?

`to_owned` is already `A generalization of Clone to borrowed data.` So you're right, I removed `.clone` now.

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 6]
```diff
@@ -14,6 +16,84 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
...
+        args.iter()
+            .map(|json| match json {
+                None => Json::None,
+                Some(json) => json.to_owned(),
+            })
+            .collect(),
+    )))
+}
+
+fn json_validator(expr: &tipb::Expr) -> Result<()> {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: json_validator is not a good name. In fact this function is only useful for JSON_OBJECT.

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 7]
```diff
@@ -6,6 +6,8 @@ use crate::codec::data_type::*;
 use crate::codec::mysql::json::*;
 use crate::Result;
 
+use std::collections::BTreeMap;
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: Please put it before line 3.

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 8]
```diff
@@ -14,6 +16,84 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
...
+}
+
+fn json_validator(expr: &tipb::Expr) -> Result<()> {
+    for chunk in expr.get_children().chunks(2) {
+        if chunk.len() == 1 {
+            return Err(other_err!(
+                "Incorrect parameter count in the call to native function 'JSON_OBJECT'"
+            ));
+        } else {
+            super::function::validate_expr_return_type(&chunk[0], Bytes::EVAL_TYPE)?;
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: EvalType::Bytes

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 9]
```diff
@@ -14,6 +16,84 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
...
+        let value = Evaluable::borrow_scalar_value_ref(&chunk[1])
+            .as_ref()
+            .map_or(Json::None, |v: &Json| v.to_owned());
+
+        pairs.insert(key, value);
+    }
+    Ok(Some(Json::Object(pairs)))
+}
+
+macro_rules! parse_opt {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: Remove this macro.

File: components/tidb_query/src/rpn_expr/impl_json.rs [hunk 10]
```diff
@@ -30,10 +110,10 @@ mod tests {
 
     use crate::rpn_expr::types::test_util::RpnFnScalarEvaluator;
 
+    use std::str::FromStr;
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_json.rs]: Put before line 111.

**Comment by breeswish** [comment]
Prefer to separate into multiple PRs. I'm fine with json_array. However the rest is not reviewed yet.

**Pull request closed without merging**